src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -4061,58 +4061,131 @@ void C2_MacroAssembler::masked_op(int ideal_opc, int mask_len, KRegister dst,
 
 #ifdef _LP64
 void C2_MacroAssembler::vector_mask_operation(int opc, Register dst, KRegister mask,
-                                              Register tmp, int masklen, int masksize,
-                                              int vec_enc) {
+                                              int masklen, int masksize, int vec_enc) {
+  assert(VM_Version::supports_popcnt() &&
```

Comment:
New instructions like `lzcntq` and `tzcntq` are used for the optimized code gen without detecting the availability.
I'm a bit worried about that.

So do all AVX512 platforms support them?
Thanks.

Reply:
Yes, you are right. I can't find concrete evidence that AVX512 implies BMI1. In addition, `VectorMaskGen` does a check for AVX3 simultaneously with a check for BMI2. So it seems safer to do the same as we did with AVX1 - 2. As a result, I refactored the reduction step into a separate function.
Thank you very much.